File: vibecheck-core/tests/fixtures/lru_cache/gemini.py
class Node:
    """Doubly-linked list node for cache entries."""

    def __init__(self, key=0, value=0):
        self.key = key
        self.value = value
        self.prev = None
        self.next = None


class Cache:
    # LRU cache with fixed capacity
    # - O(1) get and put operations
    # - doubly linked list for ordering
    # - dict for fast key lookup

    def __init__(self, cap):
        self.cap = cap
        self.items = {}
        self.head = Node()
        self.tail = Node()
        self.head.next = self.tail
        self.tail.prev = self.head

    def get(self, key):
        found = key in self.items
        node = self.items.get(key)
        value = node.value if found else -1
        if found:
            self._detach(node)
            self._attach(node)
        return value

    def put(self, key, value):
        # - update value if key exists
        # - evict tail entry when full
        # - always move to front
        if key in self.items:
            node = self.items[key]
            node.value = value
            self._detach(node)
            self._attach(node)
            return

        full = len(self.items) >= self.cap
        if full:
            old = self.tail.prev
            self._detach(old)
            del self.items[old.key]

        node = Node(key, value)
        self.items[key] = node
        self._attach(node)

    def _detach(self, node):
        # - unlink node from its neighbors
        # - stitch prev and next together
        # - clear node pointers
        prev = node.prev
        after = node.next
        prev.next = after
        after.prev = prev

    def _attach(self, node):
        first = self.head.next
        self.head.next = node
        node.prev = self.head
        node.next = first
        first.prev = node

    def size(self):
        return len(self.items)

    def peek(self, key):
        found = key in self.items
        return self.items[key].value if found else None

    def clear(self):
        self.items.clear()
        self.head.next = self.tail
        self.tail.prev = self.head

    def keys(self):
        result = []
        node = self.head.next
        while node != self.tail:
            result.append(node.key)
            node = node.next
        empty = len(result) == 0
        return result if not empty else []

File: vibecheck-core/tests/fixtures/lru_cache/gemini.rs
use std::collections::HashMap;

// LRU Cache implementation
// - fixed capacity eviction
// - O(1) get and put via HashMap + linked list
// - tracks access order for recency

struct Node {
    key: i32,
    value: i32,
    prev: Option<usize>,
    next: Option<usize>,
}

struct Cache {
    items: HashMap<i32, usize>,
    nodes: Vec<Node>,
    head: Option<usize>,
    tail: Option<usize>,
    cap: usize,
}

impl Cache {
    fn new(cap: usize) -> Self {
        Cache {
            items: HashMap::new(),
            nodes: Vec::new(),
            head: None,
            tail: None,
            cap,
        }
    }

    fn get(&mut self, key: i32) -> Option<i32> {
        let idx = *self.items.get(&key)?;
        self.detach(idx);
        self.attach(idx);
        Some(self.nodes[idx].value)
    }

    fn put(&mut self, key: i32, value: i32) {
        // - update existing entry if present
        // - evict oldest when at capacity
        // - attach new entry at head
        if let Some(&idx) = self.items.get(&key) {
            self.nodes[idx].value = value;
            self.detach(idx);
            self.attach(idx);
            return;
        }
        let evict = self.items.len() >= self.cap;
        if evict {
            let old = self.tail.unwrap();
            self.detach(old);
            self.items.remove(&self.nodes[old].key);
        }
        let idx = self.nodes.len();
        let node = Node {
            key,
            value,
            prev: None,
            next: None,
        };
        self.nodes.push(node);
        self.items.insert(key, idx);
        self.attach(idx);
    }

    fn detach(&mut self, idx: usize) {
        let prev = self.nodes[idx].prev;
        let next = self.nodes[idx].next;
        // - relink neighbors around this node
        // - update head/tail if needed
        // - clear own pointers
        match prev {
            Some(p) => self.nodes[p].next = next,
            None => self.head = next,
        }
        match next {
            Some(n) => self.nodes[n].prev = prev,
            None => self.tail = prev,
        }
        self.nodes[idx].prev = None;
        self.nodes[idx].next = None;
    }

    fn attach(&mut self, idx: usize) {
        self.nodes[idx].next = self.head;
        let had = self.head.is_some();
        if had {
            let old = self.head.unwrap();
            self.nodes[old].prev = Some(idx);
        }
        self.head = Some(idx);
        self.tail = if had { self.tail } else { Some(idx) };
    }

    fn len(&self) -> usize {
        self.items.len()
    }
}
